### ISSUE TITLE
"Improvised tools" and how to make them

### ISSUE THREAD

**Issue by GuardianDll** [CleverRaven/Cataclysm-DDA#67144]
### Is your feature request related to a problem? Please describe.

AKA tools, that are not designed for stuff they perform
It's something we discussed a lot in discord, but no one actually fill the feature requrest, so: there is a huge variety of items in the game, and actually lot of them could be purposed for tasks, they are not really designed for - using knife or scissors to screw something simple, or using bronze tools for the same tasks as steel one. One of the issues for this is the fact, that our tools is infinitely durable.

### Solution you would like.

Expand the `TOOL` `qualities` syntax, that has a field, where you can define for how durable the item would be, if you use this quality of the item. For example: the old syntax:
```json

"qualities": [ [ "CUT", 2 ], [ "CUT_FINE", 1 ], [ "BUTCHER", 22 ] ],

```

Could be advanced into:
```json

"qualities": [ [ "CUT", 2 ], [ "CUT_FINE", 1 ], [ "BUTCHER", 22 ], [ "SCREW", 1, 1 ] ],

```
which adds SCREW quality of level 1, and imply you will damage the item every time you use it like this

another example: the primitive hammer (HAMMER", 2) may look like this:

```json

"qualities": [ [ "HAMMER", 2 ], [ "HAMMER", 3, 10 ] ],

```

which imply, that stone hammer can be used as HAMMER quality 2 with no problem, but every task that uses HAMMER 3 quality has 1/10 probability to destroy an item

so `10` means 1/10 prob, 1 means `1/1` to apply damage on item, we technically can use floats here to allow `1/0.5` as "item took 2 damage for every task", but that's excessive i think. `0` imply "item would be destroyed after using it once", and default value `none` would mean the item can't be destroyed by using this tool quality

Also, the game probably need to highlight this tool in some way like yellow or light red color in the recipe, and create a popup like "Cafting this will use X, which can damage the tool.  Proceed?", but that's a minor stuff

### Additional context
I probably need to hear another thoughts if i miss something important

**Comment by Vaskritaya** [comment]
While there's probably plenty of ways to damage tools used improperly, I feel like the more likely situation is that you take longer to do a thing. Eg, I struggle to imagine damaging a knife you're using as a screwdriver on smaller screws; I've used a pocketknife as an improvised screwdriver for electronics probably dozens of times. That being said, it is more fiddly and precise than using an actual screwdriver, so I probably took longer than if I used the proper tools.

tldr: I feel like the majority of improvised tool uses affect time taken instead of tool safety. I know recipes taking variable time depending on tools used is already a highly desired mechanic for eg. power tools vs hand tools and the like, so I guess add this onto the pile of things that need better recipe infrastructure.

**Comment by GuardianDll** [comment]
Last time i tried to use my knife to unscrew something, the tip of the blade got broken really easy, because, well, kitchen knifes are not designed to screw anything, as they have no strength against side pressure.

Time is also a good idea to count here, that may work even better (or we can combine it), and even may solve the problem with duplicating recipes (like, if we want to add a sewing machine, we could simply add it, and define it has SEWING quality, that is performed 10 times faster than manual work, and it would apply to all recipes that uses sewing quality)

**Comment by GuardianDll** [comment]
I actually forgot we can nest, and it surely will make it better

**Comment by Brambor** [comment]
> Time is also a good idea to count here, that may work even better (or we can combine it), and even may solve the problem with duplicating recipes (like, if we want to add a sewing machine, we could simply add it, and define it has SEWING quality, that is performed 10 times faster than manual work, and it would apply to all recipes that uses sewing quality)

Speed seems very nice at first, but a recipe that needs sewing only 5% of the time should take 95.5% of total time, not 10% of total time with sewing machine. This would work with multi staged recipes, the holy grail of crafting.

**Comment by Brambor** [comment]
In your proposal, tools overlap, how would the speedup work then? If you use oven and a knife at the same time, does it help that you have faster oven?

**Comment by ilovehans10** [comment]
In my first suggestion, the tool speedup could happen based on the tool's speed multiplied by the time_percentage for the use of the tool in the recipe. For example, if we have an oven that offers a 2x speed multiplier for the OVEN quality then for a recipe where `{{ "id": "OVEN", "level": 1, "time_percentage": 80 }, { "id": "CUT", "level": 2, "time_percentage": 20 }}` that takes 10 seconds then the final recipe would take:

`10-(10*(80/100)*(1-1/2)+10*(20/100)*(1-1/1)) = 6 seconds`
or
`recipe_time-(recipe_time*(OVEN_time_percentage/100)*(1-1/OVEN_speed_multiplier)+recipe_time*(CUT_time_percentage/100)*(1-1/CUT_speed_multiplier)) = total time`

With my second suggestion, if you have an oven tool that has the quality of `{ "id": "OVEN", "level": 1, "automatic": TRUE }` then doing the same recipe would go as follows. You have to work on the project directly for only 20% of the craft as the 80% of the craft that is done in the oven has an oven that works automatically. So you would craft for 2 minutes and then the recipe wouldn't be done, but it will be "in the oven", so you can do some reading while you are waiting for your food. This could be implemented similarly to how tanned pelt crafting or sourdough crafting works now, but there are questions to be raised about how charges would be consumed in recipes that need them from an automatic tool.

To be clear, I know the math is janky, and the pseudocode (if you can even call it that) looks awful, but I wanted to get this idea out, and I can give more examples or reexplain as needed.

**Comment by Brambor** [comment]
Thanks, the discrepancy confused me as in your previous post time percentages were 80 + 30 = 110. Clearing that up, your explanation is reasonable.

But using two tools at the same time is also a thing, and improving one shouldn't necessarily shorten the time.

_Warning while reading the following section: I don't know much, maybe don't read it._

Although I didn't want to drag it out in this issue... What about required tools hidden under `using`?

https://github.com/CleverRaven/Cataclysm-DDA/blob/65d79079c77c3566e44a454227803eb90564f90b/data/json/recipes/armor/arms.json#L512-L523

Under `using` there are `tailoring_cotton_patchwork` and `tailoring_neoprene_patchwork` both of which are roughly:

https://github.com/CleverRaven/Cataclysm-DDA/blob/65d79079c77c3566e44a454227803eb90564f90b/data/json/requirements/tailoring.json#L272-L276

So in the original recipe, you have to specify what time percentage sewing takes. But the fact that you are sewing is within the `using` which wants to be independent of this recipe <==> in the original recipe you shouldn`t even know that you are sewing.

So would you define it in the `using` pointing to the `requirement`? Probably, I am writing this for an hour and forgot my contra argument. So I should go to sleep.

Oh! There are more tools (SEW, CUT) in each `requirement`, so... I guess in this case cutting doesn't matter and you concentrate 100% on the sewing or make it 95:5. This then has to be multiplied by the value in `using` for `tailoring_cotton_patchwork` and `tailoring_neoprene_patchwork` respectively. _I didn't write this well, but oh well..._

All in all, you are trying to fake multistaging recipes, but it would be easier to just make multistage recipes.

_To be fair, I don't know how far multistage recipes are and if hacking this together could even help bring it closer as you would research the time proportions. But then you would add complexity to code which would probably be retired older... I really don't know these things._

If this is too off-topic just mark it so.

**Comment by estebandellasilva** [comment]
But doing the change that you damage tools would lead to a Problem ... Tool selection during crafting - imagine you have several tools that have the same quality ... you would be required to choose which tool to use ... EVERY SINGLE TIME ... this adds a lot of tedium when crafting

Another Thing would be required to also be able to repair tools with the mending action  + tools should indicate that they are damaged somehow (if i am forced to use a worse hammer but i have 3 of them i want to break 1 at a time and not all 3 because the game simply does not tell me which one i am currently breaking)

Additionally ... not every recipe iss taxing for tools in the same way ... but not every recipe iss possible with every tool ( i doubt you could forge a sword with a stone hammer - i doubt that a hammer would get damaged the same during a 5 week forging session (plate mail) than during the creation of a box with nails for example (if both used the same hammer quality for example)

**Comment by ilovehans10** [comment]
> Oh! There are more tools (SEW, CUT) in each requirement, so... I guess in this case cutting doesn't matter and you concentrate 100% on the sewing or make it 95:5. This then has to be multiplied by the value in using for tailoring_cotton_patchwork and tailoring_neoprene_patchwork respectively. I didn't write this well, but oh well...

I think I get what you are say here. Is the issue that when the "using" field is what chooses the tools used then we would run into issues figuring out the percentages for the tools in requirements used? If so then we could fix this the way that you had mentioned where we give the tools inside the requirements percentages used, and also give a percentage to the requirements in the recipe itself. 

> All in all, you are trying to fake multistaging recipes, but it would be easier to just make multistage recipes.
> 
> _To be fair, I don't know how far multistage recipes are and if hacking this together could even help bring it closer as you would research the time proportions. But then you would add complexity to code which would probably be retired older... I really don't know these things._

I agree with this sentiment that this as it stands still doesn't hold a candle to multistage recipes. I feel like it is at least worth considering the fact that since the recipes will eventually split into parts that making it so that the recipes can start having time percentages added now could be helpful in the future.

> But doing the change that you damage tools would lead to a Problem ... Tool selection during crafting - imagine you have several tools that have the same quality ... you would be required to choose which tool to use ... EVERY SINGLE TIME ... this adds a lot of tedium when crafting

I don't see why you would have to choose the tool every time. It would be almost the same as it is now where if there is a tool nearby that has the quality and the quality has no damage stat (most tools won't be damaged with standard use) then the recipe will start crafting. If there isn't a tool nearby that has the quality without causing damage then you will be prompted if you really want to use a tool that will break over time.

> Another Thing would be required to also be able to repair tools with the mending action + tools should indicate that they are damaged somehow (if i am forced to use a worse hammer but i have 3 of them i want to break 1 at a time and not all 3 because the game simply does not tell me which one i am currently breaking)

I could be wrong but I am almost certain that currently tools can both be repaired, and show they are damaged already. If you are talking about them being shown as broken when picking your tool; I am pretty sure that is also already implemented. If it isn't implemented yet, then yes all of these things seem like they would be required for implementing tool damage.

It could also be nice to implement an option in the settings where you could choose if recipes would automatically use the tool closest to broken or most repaired. Possibly another option about a tool damage threshold so you can keep all of your hammers above 20% or just not use tools that can be damaged at all (the way the game currently works).

> Additionally ... not every recipe iss taxing for tools in the same way ... but not every recipe iss possible with every tool ( i doubt you could forge a sword with a stone hammer - i doubt that a hammer would get damaged the same during a 5 week forging session (plate mail) than during the creation of a box with nails for example (if both used the same hammer quality for example)

If you ask me the tool damage wouldn't all be applied at the completion of the recipe, but instead over time similar to how charges are used up on batteries. This would make it so that you would almost certainly not want to use it for making plate mail, but viable for quick making some nails.
